viam-cartographer/src/carto_facade/io.cc

```diff
@@ -0,0 +1,112 @@
+// This is an experimental integration of cartographer into RDK.
+#include "io.h"
+
+#include <pcl/io/pcd_io.h>
+#include <pcl/point_types.h>
+#include <stdio.h>
+
+#include <boost/filesystem.hpp>
+#include <fstream>  // std::ifstream
+#include <iomanip>
+#include <iostream>
+#include <string>
+
+#include "glog/logging.h"
+
+namespace viam {
+namespace carto_facade {
+namespace io {
+
+namespace fs = boost::filesystem;
+
+const std::string MakeFilenameWithTimestamp(std::string path_to_dir,
+                                            std::time_t t) {
+    char timestamp[100];
+    std::strftime(timestamp, sizeof(timestamp), time_format.c_str(),
+                  std::gmtime(&t));
+    return path_to_dir + "/" + "map_data_" + timestamp + ".pbstream";
+}
+
+cartographer::sensor::TimedPointCloudData TimedPointCloudDataFromPCDBuilder(
+    std::string file_path, double start_time) {
+    pcl::console::setVerbosityLevel(pcl::console::L_ALWAYS);
+
+    cartographer::sensor::TimedPointCloudData timed_pcd;
+    cartographer::sensor::TimedPointCloud ranges;
+
+    // Open the point cloud file
+    pcl::PointCloud<pcl::PointXYZRGB>::Ptr cloud(
+        new pcl::PointCloud<pcl::PointXYZRGB>);
+    auto err = pcl::io::loadPCDFile<pcl::PointXYZRGB>(file_path, *cloud);
+
+    if (err == -1) {
+        return timed_pcd;
+    }
+
+    double current_time = ReadTimeFromTimestamp(file_path.substr(
+        file_path.find(filename_prefix) + filename_prefix.length(),
+        file_path.find(".pcd")));
+    double time_delta = current_time - start_time;
+
+    VLOG(1) << "Accessing file " << file_path << " ... ";
+    VLOG(1) << "Loaded " << cloud->width * cloud->height << " data points";
+
+    for (size_t i = 0; i < cloud->points.size(); ++i) {
+        cartographer::sensor::TimedRangefinderPoint timed_rangefinder_point;
+        timed_rangefinder_point.position = Eigen::Vector3f(
+            cloud->points[i].x, cloud->points[i].y, cloud->points[i].z);
+        timed_rangefinder_point.time = 0 - i * 0.0001;
+
+        ranges.push_back(timed_rangefinder_point);
+    }
+
+    timed_pcd.time = cartographer::common::FromUniversal(123) +
+                     cartographer::common::FromSeconds(double(time_delta));
+    timed_pcd.origin = Eigen::Vector3f::Zero();
+    timed_pcd.ranges = ranges;
+
+    return timed_pcd;
+}
+
+double ReadTimeFromTimestamp(std::string timestamp) {
+    std::string::size_type sz;
+    auto partial_time_format = time_format.substr(0, time_format.find("."));
```

Comment:
Why do you use `auto` here?

Reply:
In Modern C++, when you don't care about the type of a variable, it is recommended to use `auto` to allow the compiler to derive the type.